pkg/member/member_add.go

```diff
@@ -0,0 +1,158 @@
+func IsMemberInCluster(logger *logrus.Logger) bool {
+	//Create etcd client
+	// TODO: Use secure transport
+	clientFactory := etcdutil.NewFactory(brtypes.EtcdConnectionConfig{
+		Endpoints:         []string{os.Getenv("ETCD_ENDPOINT")},
+		InsecureTransport: true, //TODO: is it right to use insecure transport?
+	})
+
+	// TODO: should use a retry mechanism here
+	cli, _ := clientFactory.NewCluster()
+	defer cli.Close()
+	logger.Info("Etcd client created")
+
+	// List members in cluster
+	memListCtx, cancel := context.WithTimeout(context.TODO(), brtypes.DefaultEtcdConnectionTimeout)
+	etcdMemberList, err := cli.MemberList(memListCtx)
+	defer cancel()
+	if err != nil {
+		logger.Warn("Could not list any etcd members", err)
+		return true
+	}
+
+	for _, y := range etcdMemberList.Members {
+		if y.Name == os.Getenv("POD_NAME") {
+			return true
+		}
+	}
+
+	return false
+}
+
+func getMemberURL() string {
+	end := strings.Split(os.Getenv("ETCD_ENDPOINT"), "//")
+	memberURL := end[0] + "//" + os.Getenv("POD_NAME") + "." + end[1]
+	return memberURL
+}
+
+// PromoteMember promotes an etcd member from a learner to a voting member of the cluster. This will succeed only if its logs are caught up with the leader
+func PromoteMember(ctx context.Context, logger *logrus.Entry) {
+	for {
+		// TODO: Use secure transport
+		clientFactory := etcdutil.NewFactory(brtypes.EtcdConnectionConfig{
+			Endpoints:         []string{os.Getenv("ETCD_ENDPOINT")},
+			InsecureTransport: true,
+		})
+		cli, _ := clientFactory.NewCluster()
+
+		//List all members in the etcd cluster
+		//Member URL will appear in the memberlist call response as soon as the member has been added to the cluster as a learner
+		//However, the name of the member will appear only if the member has started running
+		memListCtx, memListCtxcancel := context.WithTimeout(context.TODO(), brtypes.DefaultEtcdConnectionTimeout)
+		etcdList, memListErr := cli.MemberList(memListCtx)
+		memListCtxcancel()
+
+		if memListErr != nil {
+			logger.Info("error listing members: ", memListErr)
+			cli.Close()
+			continue
+		}
+
+		//TODO: Simplify logic below
+		var promoted bool
+		promoted = false
+		for _, y := range etcdList.Members {
+			if y.Name == os.Getenv("POD_NAME") {
+				logger.Info("Promoting member ", y.Name)
+				memPromoteCtx, cancel := context.WithTimeout(context.TODO(), brtypes.DefaultEtcdConnectionTimeout)
+				//Member promote call will succeed only if member is in sync with leader, and will error out otherwise
+				_, memPromoteErr := cli.MemberPromote(memPromoteCtx, y.ID)
+				cancel()
+				if memPromoteErr == nil {
+					//Exit if member is successfully promoted or if member is not a learner
+					promoted = true
+					logger.Info("Member promoted ", y.Name, " : ", y.ID)
+					break
+				}
+				if strings.Contains(rpctypes.ErrGRPCMemberNotLearner.Error(), memPromoteErr.Error()) {
+					//Exit if member is already part of the cluster
```

Comment:
If there is an error why do you have to loop through the rest of the members? The intent of this method is to only promote one member